File: springmvc/src/main/java/io/github/gaplo917/springmvc/services/IOService.java
package io.github.gaplo917.springmvc.services;

import io.github.gaplo917.springmvc.data.DummyResponse;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.concurrent.*;

@Service
public class IOService {
  private ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();

  public DummyResponse blockingIO(Long ioDelay) {
    try {
      Thread.sleep(ioDelay);
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
    return DummyResponse.dummy(null);
  }

  public List<DummyResponse> dependentBlockingIO(Long ioDelay, DummyResponse resp) {
    try {
      Thread.sleep(ioDelay);
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
    return List.of(resp, DummyResponse.dummy(null));
  }

  public CompletableFuture<DummyResponse> nonBlockingIO(Long ioDelay) {
    return CompletableFuture.supplyAsync(
        () -> DummyResponse.dummy(null),
        CompletableFuture.delayedExecutor(ioDelay, TimeUnit.MILLISECONDS, scheduledExecutorService)
    );
  }

  public CompletableFuture<List<DummyResponse>> dependentNonBlockingIO(Long ioDelay, DummyResponse resp) {
    return CompletableFuture.supplyAsync(
        () -> List.of(resp, DummyResponse.dummy(null)),
        CompletableFuture.delayedExecutor(ioDelay, TimeUnit.MILLISECONDS, scheduledExecutorService)
    );
  }

}

File: frameworks/aspnet6/Services/IOService.cs
namespace aspnet6.Services;

public class IoService
{
    public DummyResponse HandleIo(int delay)
    {
        Thread.Sleep(delay);
        return DummyResponse.Dummy();
    }

    public List<DummyResponse> HandleDependentIo(int delay, DummyResponse response)
    {
        Thread.Sleep(delay);
        return new List<DummyResponse> { response, DummyResponse.Dummy() };
    }

    public async Task<DummyResponse> HandleIoAsync(int delay)
    {
        await Task.Delay(delay);
        return DummyResponse.Dummy();
    }

    public async Task<List<DummyResponse>> HandleDependentIoAsync(int delay, DummyResponse response)
    {
        await Task.Delay(delay);
        return new List<DummyResponse> { response, DummyResponse.Dummy() };
    }
}